File: concepts/06_text_binary_and_serialization/05_serialization_clone_and_transfer/lua/test_02_numeric_and_trust_boundaries.lua
-- Common question: where do numeric fidelity, versioning, and trust constrain transfer formats?
-- Inputs: 64-bit integers, doubles, dumped bytecode, text-only loading, and hostile source.
-- Observations: exact integer packing, floating precision, binary mode gates, and executable-code risk.
-- polyglot-family: text_binary_and_serialization
-- polyglot-concept: serialization_clone_and_transfer
-- polyglot-related: languages/lua/tooling_and_runtime/14_standalone_and_bytecode/
-- polyglot-related+: test_108_bytecode_compilation_and_loading.lua

local t = require("support.assertions")

local integer = math.maxinteger
local bytes = string.pack("<jdn", integer, 0.1, 0.1)
local decoded_integer, double_value, native_value = string.unpack("<jdn", bytes)
t.equal(decoded_integer, integer)
t.equal(double_value, 0.1)
t.equal(native_value, 0.1)

local bytecode = string.dump(function() return 42 end, true)
t.equal(assert(load(bytecode, "trusted-local", "b"))(), 42)
t.equal(select(1, load(bytecode, "text-only", "t")), nil)

local marker = {value = false}
local source = "return function(mark) mark.value = true end"
assert(load(source, "source", "t"))()(marker)
t.truth(marker.value)

-- Source and bytecode are executable, not safe data formats; bytecode also lacks cross-version portability.
t.done()

File: concepts/06_text_binary_and_serialization/05_serialization_clone_and_transfer/ruby/test_02_numeric_and_trust_boundaries.rb
# 共同问题：数值精度、版本和信任边界怎样限制传输格式。
# 输入：大整数、浮点、Marshal header、JSON 非数值和不可信 YAML tag；观察：保真度、拒绝策略及版本。
# polyglot-family: text_binary_and_serialization
# polyglot-concept: serialization_clone_and_transfer
# polyglot-related: languages/ruby/standard_library/10_data_time_and_text/test_074_yaml_and_marshal_trust_boundaries.rb

require "assertions"
require "json"
require "yaml"

A = PolyglotAssertions

large = 2**100
A.equal(large, JSON.parse(JSON.generate(large)))
A.near(Math::PI, JSON.parse(JSON.generate(Math::PI)))
A.raises(JSON::GeneratorError) { JSON.generate(Float::NAN) }

dump = Marshal.dump({value: large})
A.equal(Marshal::MAJOR_VERSION, dump.getbyte(0))
A.equal(Marshal::MINOR_VERSION, dump.getbyte(1))
A.equal({value: large}, Marshal.load(dump))

unsafe_yaml = "--- !ruby/object:Object {}\n"
A.raises(Psych::DisallowedClass) { YAML.safe_load(unsafe_yaml, permitted_classes: []) }

A.done
